Publish cp313t (free-threaded) wheels

Currently `pip install libcst` will do a from-source build on free-threaded 3.13.0:

```
goldbaum at Nathans-MBP in ~/Documents/cryptography on main!
± pip install libcst
Collecting libcst
  Downloading libcst-1.5.0.tar.gz (772 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 773.0/773.0 kB 22.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting pyyaml>=5.2 (from libcst)
  Using cached PyYAML-6.0.2-cp313-cp313t-macosx_14_0_arm64.whl
Building wheels for collected packages: libcst
  Building wheel for libcst (pyproject.toml) ... done
  Created wheel for libcst: filename=libcst-1.5.0-cp313-cp313t-macosx_14_0_arm64.whl size=2027094 sha256=ec7f1a08268640717734e28b8454ca23043112fb2f002b46694494c82cf6b6bf
  Stored in directory: /Users/goldbaum/Library/Caches/pip/wheels/af/db/82/f05ed1829e60f978a9af19520c78cb52df1bed762e579d7f84
Successfully built libcst
Installing collected packages: pyyaml, libcst
Successfully installed libcst-1.5.0 pyyaml-6.0.2

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: pip install --upgrade pip

goldbaum at Nathans-MBP in ~/Documents/cryptography on main!
± python -VV
Python 3.13.0 experimental free-threading build (main, Nov  5 2024, 16:45:19) [Clang 16.0.0 (clang-1600.0.26.3)]
```

The build succeeds, but it probably shouldn't, and immediately segfaults if I try to import anything that imports the native module:

```
$  python
Python 3.13.0 experimental free-threading build (main, Nov  5 2024, 16:45:19) [Clang 16.0.0 (clang-1600.0.26.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import libcst.codegen.gen_matcher_classes as matcher_codegen
[1]    33127 segmentation fault  python
```

I think the first step is likely going to be updating the PyO3 dependency to 0.23 once that comes out (hopefully this week). Since the crate is at PyO3 0.20 right now, updating to 0.21 and 0.22 first will likely help.

Thanks! After building with that PR I'm able to get some more useful compiler errors and warnings:

<details>

```
warning: unused imports: `DeflatedAsynchronous as Asynchronous` and `DeflatedFormattedStringText as FormattedStringText`
  --> libcst/src/nodes/mod.rs:55:29
   |
55 | ...   DeflatedArg as Arg, DeflatedAsynchronous as Asynchronous, Deflated...
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
66 | ...   DeflatedFormattedStringText as FormattedStringText, DeflatedFrom a...
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `DeflatedImportStar as ImportStar`
  --> libcst/src/nodes/mod.rs:89:9
   |
89 |         DeflatedImportStar as ImportStar, DeflatedSemicolon as Semicolon,
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: use of deprecated trait `pyo3::IntoPy`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
   --> libcst/src/nodes/traits.rs:121:46
    |
121 |     use pyo3::{types::PyAny, types::PyTuple, IntoPy, PyObject, PyResult...
    |                                              ^^^^^^
    |
    = note: `#[warn(deprecated)]` on by default

warning: use of deprecated trait `pyo3::IntoPy`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
  --> libcst/src/parser/errors.rs:32:16
   |
32 |     use pyo3::{IntoPy, PyErr, PyErrArguments, Python};
   |                ^^^^^^

warning: use of deprecated associated function `pyo3::types::PyModule::import_bound`: renamed to `PyModule::import`
    --> libcst/src/nodes/expression.rs:2539:44
     |
2539 |                     let libcst = PyModule::import_bound(py, "libcst")?;
     |                                            ^^^^^^^^^^^^

warning: use of deprecated method `pyo3::types::IntoPyDict::into_py_dict_bound`: renamed to `IntoPyDict::into_py_dict`
    --> libcst/src/nodes/expression.rs:2551:22
     |
2551 |                     .into_py_dict_bound(py);
     |                      ^^^^^^^^^^^^^^^^^^

warning: use of deprecated associated function `pyo3::types::PyModule::import_bound`: renamed to `PyModule::import`
    --> libcst/src/nodes/expression.rs:2575:44
     |
2575 |                     let libcst = PyModule::import_bound(py, "libcst")?;
     |                                            ^^^^^^^^^^^^

warning: use of deprecated method `pyo3::types::IntoPyDict::into_py_dict_bound`: renamed to `IntoPyDict::into_py_dict`
    --> libcst/src/nodes/expression.rs:2596:22
     |
2596 |                     .into_py_dict_bound(py);
     |                      ^^^^^^^^^^^^^^^^^^

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
   --> libcst/src/nodes/traits.rs:138:21
    |
138 |             Ok(self.into_py(py))
    |                     ^^^^^^^

warning: use of deprecated associated function `pyo3::types::PyTuple::new_bound`: renamed to `PyTuple::new`
   --> libcst/src/nodes/traits.rs:173:25
    |
173 |             Ok(PyTuple::new_bound(py, converted).into())
    |                         ^^^^^^^^^

error[E0599]: no method named `into_py` found for struct `PyTuple` in the current scope
   --> libcst/src/nodes/traits.rs:179:21
    |
179 |             Ok(self.into_py(py))
    |                     ^^^^^^^
    |
help: there is a method `into` with a similar name, but with different arguments
   --> /Users/goldbaum/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:450:5
    |
450 |     fn into(self) -> T;
    |     ^^^^^^^^^^^^^^^^^^^

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
   --> libcst/src/nodes/traits.rs:185:21
    |
185 |             Ok(self.into_py(py))
    |                     ^^^^^^^

error[E0599]: no method named `into_py` found for reference `&T` in the current scope
   --> libcst/src/nodes/traits.rs:194:21
    |
194 |             Ok(self.into_py(py))
    |                     ^^^^^^^
    |
    = help: items from traits can only be used if the type parameter is bounded by the trait
help: there is a method `into` with a similar name, but with different arguments
   --> /Users/goldbaum/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:450:5
    |
450 |     fn into(self) -> T;
    |     ^^^^^^^^^^^^^^^^^^^
help: the following trait defines an item `into_py`, perhaps you need to restrict type parameter `T` with it:
    |
191 |         T: AsRef<PyAny> + IntoPy,
    |                         ++++++++

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
  --> libcst/src/parser/errors.rs:63:47
   |
63 |                     ("message", e.to_string().into_py(py)),
   |                                               ^^^^^^^

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
  --> libcst/src/parser/errors.rs:64:37
   |
64 |                     ("lines", lines.into_py(py)),
   |                                     ^^^^^^^

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
  --> libcst/src/parser/errors.rs:65:45
   |
65 |                     ("raw_line", (line + 1).into_py(py)),
   |                                             ^^^^^^^

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
  --> libcst/src/parser/errors.rs:66:40
   |
66 |                     ("raw_column", col.into_py(py)),
   |                                        ^^^^^^^

warning: use of deprecated method `pyo3::types::IntoPyDict::into_py_dict_bound`: renamed to `IntoPyDict::into_py_dict`
  --> libcst/src/parser/errors.rs:68:18
   |
68 |                 .into_py_dict_bound(py);
   |                  ^^^^^^^^^^^^^^^^^^

warning: use of deprecated associated function `pyo3::types::PyModule::import_bound`: renamed to `PyModule::import`
  --> libcst/src/parser/errors.rs:70:31
   |
70 | ...   PyModule::import_bound(py, "libcst").expect("libcst cannot be impo...
   |                 ^^^^^^^^^^^^

warning: use of deprecated associated function `pyo3::PyErr::from_value_bound`: renamed to `PyErr::from_value`
  --> libcst/src/parser/errors.rs:71:24
   |
71 |                 PyErr::from_value_bound(
   |                        ^^^^^^^^^^^^^^^^

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
  --> libcst/src/parser/errors.rs:85:42
   |
85 |                 ("message", self.message.into_py(py)),
   |                                          ^^^^^^^

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
  --> libcst/src/parser/errors.rs:86:38
   |
86 |                 ("lines", self.lines.into_py(py)),
   |                                      ^^^^^^^

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
  --> libcst/src/parser/errors.rs:87:44
   |
87 |                 ("raw_line", self.raw_line.into_py(py)),
   |                                            ^^^^^^^

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
  --> libcst/src/parser/errors.rs:88:48
   |
88 |                 ("raw_column", self.raw_column.into_py(py)),
   |                                                ^^^^^^^

warning: use of deprecated method `pyo3::types::IntoPyDict::into_py_dict_bound`: renamed to `IntoPyDict::into_py_dict`
  --> libcst/src/parser/errors.rs:90:14
   |
90 |             .into_py_dict_bound(py)
   |              ^^^^^^^^^^^^^^^^^^

warning: use of deprecated method `pyo3::IntoPy::into_py`: `IntoPy` is going to be replaced by `IntoPyObject`. See the migration guide (https://pyo3.rs/v0.23/migration) for more information.
  --> libcst/src/parser/errors.rs:91:14
   |
91 |             .into_py(py)
   |              ^^^^^^^

For more information about this error, try `rustc --explain E0599`.
warning: `libcst` (lib) generated 24 warnings
```

</details>

Anyone trying to do this with a free-threaded Python may run into the fact that both [`pip install hatch`](https://github.com/pypa/hatch/issues/1801) and [`pip install jupyter`](https://github.com/jupyterlab/jupyterlab/issues/16915) are broken right now for free-threaded Python, so you'll need to install a slightly hacked version of hatch that doesn't bring in `python-zstandard` at install time and comment out `jupyter` and `nbsphinx` in the libcst `pyproject.toml`.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -13,8 +13,13 @@ jobs:
       fail-fast: false
       matrix:
         os: [macos-latest, ubuntu-latest, windows-latest]
-        python-version: ["3.9", "3.10", "3.11", "3.12", "3.13"]
+        python-version: ["3.9", "3.10", "3.11", "3.12", "3.13", "3.13t"]
     steps:
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.10"
+      - name: Install hatch
+        run: pip install -U hatch
       - uses: actions/checkout@v4
         with:
           fetch-depth: 0
@@ -24,9 +29,6 @@ jobs:
           cache: pip
           cache-dependency-path: "pyproject.toml"
           python-version: ${{ matrix.python-version }}
-      - name: Install hatch
-        run: |
-          pip install -U hatch
       - uses: actions-rs/toolchain@v1
         with:
           toolchain: stable
@@ -43,50 +45,6 @@ jobs:
         run: |
           hatch run coverage combine .coverage.pure
           hatch run coverage report
-
-  # TODO:
-  # merge into regular CI once hatch has support for creating environments on
-  # the free-threaded build: https://github.com/pypa/hatch/issues/1931
-  free-threaded-tests:
-    name: "test (${{ matrix.os }}, 3.13t)"
-    runs-on: ${{ matrix.os }}
-    strategy:
-      fail-fast: false
-      matrix:
-        os: [macos-latest, ubuntu-latest, windows-latest]
-    steps:
-      - uses: actions/checkout@v4
-        with:
-          fetch-depth: 0
-          persist-credentials: false
-      - uses: actions/setup-python@v5
-        with:
-          cache: pip
-          cache-dependency-path: "pyproject.toml"
-          python-version: '3.13t'
-      - name: Build LibCST
-        run: |
-          # Install build-system.requires dependencies
-          pip install setuptools setuptools-scm setuptools-rust wheel
-          # Jupyter is annoying to install on free-threaded Python
-          pip install -e .[dev-without-jupyter]
-      - name: Native Parser Tests
-        # TODO: remove when native modules declare free-threaded support
-        env:
-          PYTHON_GIL: '0'
-        run: |
-          python -m coverage run -m libcst.tests
-      - name: Pure Parser Tests
-        env:
-          COVERAGE_FILE: .coverage.pure
-          LIBCST_PARSER_TYPE: pure
-        run: |
-          python -m coverage run -m libcst.tests
-      - name: Coverage
-        run: |
-          python -m coverage combine .coverage.pure
-          python -m coverage report
-
 
   # Run linters
   lint:
@@ -139,7 +97,7 @@ jobs:
       - name: Install hatch
         run: pip install -U hatch
       - uses: ts-graphviz/setup-graphviz@v2
-      - run: hatch run docs
+      - run: hatch run docs:docs
       - name: Archive Docs
         uses: actions/upload-artifact@v4
         with:

--- a/libcst/codemod/_cli.py
+++ b/libcst/codemod/_cli.py
@@ -8,18 +8,19 @@ Provides helpers for CLI interaction.
 """
 
 import difflib
+import functools
 import os.path
 import re
 import subprocess
 import sys
 import time
 import traceback
-from concurrent.futures import as_completed, Executor, ProcessPoolExecutor
+from concurrent.futures import as_completed, Executor
 from copy import deepcopy
 from dataclasses import dataclass
 from multiprocessing import cpu_count
 from pathlib import Path
-from typing import AnyStr, cast, Dict, List, Optional, Sequence, Type, Union
+from typing import AnyStr, Callable, cast, Dict, List, Optional, Sequence, Type, Union
 from warnings import warn
 
 from libcst import parse_module, PartialParserConfig
@@ -624,14 +625,20 @@ def parallel_exec_transform_with_prettyprint(  # noqa: C901
         python_version=python_version,
     )
 
-    pool_impl: type[Executor]
+    pool_impl: Callable[[], Executor]
     if total == 1 or jobs == 1:
         # Simple case, we should not pay for process overhead.
         # Let's just use a dummy synchronous executor.
         jobs = 1
         pool_impl = DummyExecutor
+    elif getattr(sys, "_is_gil_enabled", lambda: False)():  # pyre-ignore[16]
+        from concurrent.futures import ThreadPoolExecutor
+
+        pool_impl = functools.partial(ThreadPoolExecutor, max_workers=jobs)
     else:
-        pool_impl = ProcessPoolExecutor
+        from concurrent.futures import ProcessPoolExecutor
+
+        pool_impl = functools.partial(ProcessPoolExecutor, max_workers=jobs)
         # Warm the parser, pre-fork.
         parse_module(
             "",
@@ -650,7 +657,7 @@ def parallel_exec_transform_with_prettyprint(  # noqa: C901
         deepcopy(transform.context.scratch) if isinstance(transform, Codemod) else {}
     )
 
-    with pool_impl(max_workers=jobs) as executor:  # type: ignore
+    with pool_impl() as executor:  # type: ignore
         try:
             futures = [
                 executor.submit(

--- a/libcst/codemod/_dummy_pool.py
+++ b/libcst/codemod/_dummy_pool.py
@@ -22,9 +22,6 @@ class DummyExecutor(Executor):
     Synchronous dummy `concurrent.futures.Executor` analogue.
     """
 
-    def __init__(self, max_workers: Optional[int] = None) -> None:
-        pass
-
     def submit(
         self,
         fn: Callable[Params, Return],

--- a/native/libcst/src/py.rs
+++ b/native/libcst/src/py.rs
@@ -6,7 +6,7 @@
 use crate::nodes::traits::py::TryIntoPy;
 use pyo3::prelude::*;
 
-#[pymodule]
+#[pymodule(gil_used = false)]
 #[pyo3(name = "native")]
 pub fn libcst_native(_py: Python, m: &Bound<PyModule>) -> PyResult<()> {
     #[pyfn(m)]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -98,6 +98,7 @@ skip = [
     "*-musllinux_s390x",
     "*-musllinux_armv7l",
 ]
+enable = ["cpython-freethreading"]
 
 [tool.cibuildwheel.linux]
 environment-pass = ["LIBCST_NO_LOCAL_SCHEME"]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -18,33 +18,9 @@ classifiers = [
     "Typing :: Typed",
 ]
 requires-python = ">=3.9"
-dependencies = ["pyyaml>=5.2"]
-
-[project.optional-dependencies]
-dev = [
-    "libcst[dev-without-jupyter]",
-    "jupyter>=1.0.0",
-    "nbsphinx>=0.4.2",
-]
-dev-without-jupyter = [
-    "black==25.1.0",
-    "coverage[toml]>=4.5.4",
-    "build>=0.10.0",
-    "fixit==2.1.0",
-    "flake8==7.2.0",
-    "Sphinx>=5.1.1",
-    "hypothesis>=4.36.0",
-    "hypothesmith>=0.0.4",
-    "maturin>=1.7.0,<1.8",
-    "prompt-toolkit>=2.0.9",
-    "pyre-check==0.9.18; platform_system != 'Windows'",
-    "setuptools_scm>=6.0.1",
-    "sphinx-rtd-theme>=0.4.3",
-    "ufmt==2.8.0",
-    "usort==1.0.8.post1",
-    "setuptools-rust>=1.5.2",
-    "slotscheck>=0.7.1",
-    "jinja2==3.1.6",
+dependencies = [
+    "pyyaml>=5.2; python_version < '3.13'",
+    "pyyaml-ft; python_version >= '3.13'",
 ]
 
 [project.urls]
@@ -63,10 +39,26 @@ show_missing = true
 skip_covered = true
 
 [tool.hatch.envs.default]
-features = ["dev"]
+installer = "uv"
+dependencies = [
+    "black==25.1.0",
+    "coverage[toml]>=4.5.4",
+    "build>=0.10.0",
+    "fixit==2.1.0",
+    "flake8==7.2.0",
+    "hypothesis>=4.36.0",
+    "hypothesmith>=0.0.4",
+    "maturin>=1.7.0,<1.8",
+    "prompt-toolkit>=2.0.9",
+    "pyre-check==0.9.18; platform_system != 'Windows'",
+    "setuptools_scm>=6.0.1",
+    "ufmt==2.8.0",
+    "usort==1.0.8.post1",
+    "setuptools-rust>=1.5.2",
+    "slotscheck>=0.7.1",
+]
 
 [tool.hatch.envs.default.scripts]
-docs = "sphinx-build -ab html docs/source docs/build"
 fixtures = ["python scripts/regenerate-fixtures.py", "git diff --exit-code"]
 format = "ufmt format libcst scripts"
 lint = [
@@ -77,6 +69,17 @@ lint = [
 ]
 test = ["python --version", "python -m coverage run -m libcst.tests"]
 typecheck = ["pyre --version", "pyre check"]
+
+[tool.hatch.envs.docs]
+extra-dependencies = [
+    "Sphinx>=5.1.1",
+    "sphinx-rtd-theme>=0.4.3",
+    "jupyter>=1.0.0",
+    "nbsphinx>=0.4.2",
+    "jinja2==3.1.6",
+]
+[tool.hatch.envs.docs.scripts]
+docs = "sphinx-build -ab html docs/source docs/build"
 
 [tool.slotscheck]
 exclude-modules = '^libcst\.(testing|tests)'

